File: packages/server/src/socket/memory.ts
import { eq, asc } from "drizzle-orm";
import { db } from "../db";
import { storyQueue as storyQueueTable, sessions } from "../db/schema";
import { generateId } from "../lib/id";
import type { Participant, ParticipantRole, PlannedStory } from "../types";

interface SocketMeta {
  roomId: string;
  name: string;
  email: string;
  isModerator: boolean;
  role: ParticipantRole;
}

// socketId → metadata del socket
const socketMap = new Map<string, SocketMeta>();

// roomId → participantes (por nombre para dedup en reconexión)
const roomParticipants = new Map<string, Map<string, Participant>>();

export function addParticipant(
  socketId: string,
  roomId: string,
  name: string,
  email: string,
  isModerator: boolean,
  role: ParticipantRole
): void {
  socketMap.set(socketId, { roomId, name, email, isModerator, role });

  if (!roomParticipants.has(roomId)) {
    roomParticipants.set(roomId, new Map());
  }
  roomParticipants.get(roomId)!.set(name, { name, isModerator, hasVoted: false, role });
}

export function removeParticipant(socketId: string): SocketMeta | undefined {
  const meta = socketMap.get(socketId);
  if (!meta) return undefined;

  socketMap.delete(socketId);

  // Solo eliminar si no hay otro socket con el mismo nombre en la sala
  const otherSocket = [...socketMap.values()].find(
    (m) => m.roomId === meta.roomId && m.name === meta.name
  );
  if (!otherSocket) {
    roomParticipants.get(meta.roomId)?.delete(meta.name);
  }

  return meta;
}

export function getParticipants(roomId: string): Participant[] {
  return [...(roomParticipants.get(roomId)?.values() ?? [])];
}

export function markVoted(roomId: string, name: string): void {
  const p = roomParticipants.get(roomId)?.get(name);
  if (p) p.hasVoted = true;
}

export function resetVotes(roomId: string): void {
  roomParticipants.get(roomId)?.forEach((p) => {
    p.hasVoted = false;
  });
}

export function getMeta(socketId: string): SocketMeta | undefined {
  return socketMap.get(socketId);
}

// Queue management — persisted in SQLite
export function getQueue(roomId: string): PlannedStory[] {
  const rows = db
    .select()
    .from(storyQueueTable)
    .where(eq(storyQueueTable.roomId, roomId))
    .orderBy(asc(storyQueueTable.position))
    .all();

  // IDs de items de la cola que ya tienen sesión con resultado guardado
  const votedIds = new Set(
    db
      .select()
      .from(sessions)
      .where(eq(sessions.roomId, roomId))
      .all()
      .filter((s) => s.result !== null && s.storyQueueId !== null)
      .map((s) => s.storyQueueId as string)
  );

  return rows.map((row) => ({
    id: row.id,
    storyName: row.storyName,
    jiraKey: row.jiraKey ?? undefined,
    isVoted: votedIds.has(row.id),
  }));
}

export function clearQueue(roomId: string): void {
  db.delete(storyQueueTable).where(eq(storyQueueTable.roomId, roomId)).run();
}

export function addToQueue(roomId: string, storyName: string, jiraKey?: string): PlannedStory {
  const existing = db
    .select({ position: storyQueueTable.position })
    .from(storyQueueTable)
    .where(eq(storyQueueTable.roomId, roomId))
    .orderBy(asc(storyQueueTable.position))
    .all();
  const maxPosition = existing.length > 0 ? Math.max(...existing.map((r) => r.position)) : 0;

  const story: PlannedStory = { id: generateId(), storyName, jiraKey };
  db.insert(storyQueueTable)
    .values({
      id: story.id,
      roomId,
      storyName,
      jiraKey: jiraKey ?? null,
      position: maxPosition + 1,
      createdAt: Date.now(),
    })
    .run();
  return story;
}

export function removeFromQueue(roomId: string, storyId: string): boolean {
  const row = db
    .select()
    .from(storyQueueTable)
    .where(eq(storyQueueTable.id, storyId))
    .get();
  if (!row || row.roomId !== roomId) return false;
  db.delete(storyQueueTable).where(eq(storyQueueTable.id, storyId)).run();
  return true;
}

export function getFromQueue(roomId: string, storyId: string): PlannedStory | undefined {
  const row = db
    .select()
    .from(storyQueueTable)
    .where(eq(storyQueueTable.id, storyId))
    .get();
  if (!row || row.roomId !== roomId) return undefined;
  return { id: row.id, storyName: row.storyName, jiraKey: row.jiraKey ?? undefined };
}

export function popFromQueue(roomId: string, storyId: string): PlannedStory | undefined {
  const row = db
    .select()
    .from(storyQueueTable)
    .where(eq(storyQueueTable.id, storyId))
    .get();
  if (!row || row.roomId !== roomId) return undefined;
  db.delete(storyQueueTable).where(eq(storyQueueTable.id, storyId)).run();
  return { id: row.id, storyName: row.storyName, jiraKey: row.jiraKey ?? undefined };
}

File: packages/server-python/memory.py
from db import fetchall, fetchone, execute, generate_id, now_ms

# socket_id → metadata del socket
_socket_map: dict[str, dict] = {}

# room_id → { name → participante }
_room_participants: dict[str, dict[str, dict]] = {}


def add_participant(
    socket_id: str,
    room_id: str,
    name: str,
    email: str,
    is_moderator: bool,
    role: str,
) -> None:
    _socket_map[socket_id] = {
        "roomId": room_id,
        "name": name,
        "email": email,
        "isModerator": is_moderator,
        "role": role,
    }
    if room_id not in _room_participants:
        _room_participants[room_id] = {}
    _room_participants[room_id][name] = {
        "name": name,
        "isModerator": is_moderator,
        "hasVoted": False,
        "role": role,
    }


def remove_participant(socket_id: str) -> dict | None:
    meta = _socket_map.pop(socket_id, None)
    if not meta:
        return None
    # Solo eliminar si no hay otro socket con el mismo nombre en la sala
    other = any(
        m["roomId"] == meta["roomId"] and m["name"] == meta["name"]
        for m in _socket_map.values()
    )
    if not other:
        _room_participants.get(meta["roomId"], {}).pop(meta["name"], None)
    return meta


def get_participants(room_id: str) -> list[dict]:
    return list(_room_participants.get(room_id, {}).values())


def mark_voted(room_id: str, name: str) -> None:
    p = _room_participants.get(room_id, {}).get(name)
    if p:
        p["hasVoted"] = True


def reset_votes(room_id: str) -> None:
    for p in _room_participants.get(room_id, {}).values():
        p["hasVoted"] = False


def get_meta(socket_id: str) -> dict | None:
    return _socket_map.get(socket_id)


# --- Gestión de cola (persistida en SQLite) ---

def get_queue(room_id: str) -> list[dict]:
    rows = fetchall(
        "SELECT * FROM story_queue WHERE room_id = ? ORDER BY position ASC",
        (room_id,),
    )
    voted_sessions = fetchall(
        "SELECT story_queue_id FROM sessions "
        "WHERE room_id = ? AND result IS NOT NULL AND story_queue_id IS NOT NULL",
        (room_id,),
    )
    voted_ids = {s["story_queue_id"] for s in voted_sessions}
    return [
        {
            "id": r["id"],
            "storyName": r["story_name"],
            "jiraKey": r["jira_key"],
            "isVoted": r["id"] in voted_ids,
        }
        for r in rows
    ]


def add_to_queue(room_id: str, story_name: str, jira_key: str | None = None) -> dict:
    rows = fetchall(
        "SELECT position FROM story_queue WHERE room_id = ? ORDER BY position ASC",
        (room_id,),
    )
    max_pos = max((r["position"] for r in rows), default=0)
    story_id = generate_id()
    execute(
        "INSERT INTO story_queue (id, room_id, story_name, jira_key, position, created_at) "
        "VALUES (?, ?, ?, ?, ?, ?)",
        (story_id, room_id, story_name, jira_key, max_pos + 1, now_ms()),
    )
    return {"id": story_id, "storyName": story_name, "jiraKey": jira_key}


def remove_from_queue(room_id: str, story_id: str) -> bool:
    row = fetchone("SELECT * FROM story_queue WHERE id = ?", (story_id,))
    if not row or row["room_id"] != room_id:
        return False
    execute("DELETE FROM story_queue WHERE id = ?", (story_id,))
    return True


def get_from_queue(room_id: str, story_id: str) -> dict | None:
    row = fetchone("SELECT * FROM story_queue WHERE id = ?", (story_id,))
    if not row or row["room_id"] != room_id:
        return None
    return {"id": row["id"], "storyName": row["story_name"], "jiraKey": row["jira_key"]}


def clear_queue(room_id: str) -> None:
    execute("DELETE FROM story_queue WHERE room_id = ?", (room_id,))
